connor/connor.go

```diff
@@ -0,0 +1,231 @@
+package connor
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"fmt"
+	"time"
+
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/sonm-io/core/proto"
+	"github.com/sonm-io/core/util"
+	"github.com/sonm-io/core/util/xgrpc"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc/credentials"
+
+	"github.com/noxiouz/zapctx/ctxlog"
+	"github.com/sonm-io/core/connor/database"
+	"github.com/sonm-io/core/connor/watchers"
+	"go.uber.org/zap"
+)
+
+const (
+	coinMarketCapTicker     = "https://api.coinmarketcap.com/v1/ticker/"
+	coinMarketCapSonmTicker = coinMarketCapTicker + "sonm/"
+	cryptoCompareCoinData   = "https://www.cryptocompare.com/api/data/coinsnapshotfullbyid/?id="
+	poolReportedHashRate    = "https://api.nanopool.org/v1/eth/reportedhashrates/"
+	poolAverageHashRate     = "https://api.nanopool.org/v1/eth/avghashrateworkers/"
+)
+
+type Connor struct {
+	key          *ecdsa.PrivateKey
+	Market       sonm.MarketClient
+	TaskClient   sonm.TaskManagementClient
+	DealClient   sonm.DealManagementClient
+	TokenClient  sonm.TokenManagementClient
+	MasterClient sonm.MasterManagementClient
+
+	cfg    *Config
+	db     *database.Database
+	logger *zap.Logger
+}
+
+func NewConnor(ctx context.Context, key *ecdsa.PrivateKey, cfg *Config) (*Connor, error) {
+	connor := &Connor{
+		key: key,
+		cfg: cfg,
+	}
+
+	creds, err := newCredentials(ctx, key)
+	if err != nil {
+		return nil, fmt.Errorf("can't create TLS credentials: %v", err)
+	}
+
+	nodeCC, err := xgrpc.NewClient(ctx, cfg.Market.Endpoint, creds)
+	if err != nil {
+		return nil, fmt.Errorf("can't create node connection: %v\r\n", err)
+	}
+
+	connor.Market = sonm.NewMarketClient(nodeCC)
+	connor.TaskClient = sonm.NewTaskManagementClient(nodeCC)
+	connor.DealClient = sonm.NewDealManagementClient(nodeCC)
+	connor.TokenClient = sonm.NewTokenManagementClient(nodeCC)
+	connor.MasterClient = sonm.NewMasterManagementClient(nodeCC)
+
+	connor.db, err = database.NewDatabaseConnect(connor.cfg.Database.Driver, connor.cfg.Database.DataSource)
+	if err != nil {
+		return nil, err
+	}
+
+	balanceReply, err := connor.TokenClient.Balance(ctx, &sonm.Empty{})
+	if err != nil {
+		connor.logger.Error("cannot load balanceReply", zap.Error(err))
+		return nil, err
+	}
+
+	connor.logger = ctxlog.GetLogger(ctx)
+	connor.logger.Info("config",
+		zap.String("eth node address", cfg.Market.Endpoint),
+		zap.String("key", crypto.PubkeyToAddress(key.PublicKey).String()))
+	connor.logger.Info("balance",
+		zap.String("live", balanceReply.GetLiveBalance().Unwrap().String()),
+		zap.String("Side", balanceReply.GetSideBalance().ToPriceString()))
+	connor.logger.Info("configuring connor", zap.Any("config", cfg))
+	return connor, nil
+}
+
+func (c *Connor) Serve(ctx context.Context) error {
```

Comment:
I have got no idea what's happeninig in this function, neither will you in two months time at most. Moving the switch cases out of the body of the function would be a great start to improve readability.